x/wasm/internal/keeper/ioutil.go

```diff
@@ -28,6 +30,24 @@ func uncompress(src []byte) ([]byte, error) {
 		return nil, err
 	}
 	zr.Multistream(false)
+	defer zr.Close()
+	return ioutil.ReadAll(LimitReader(zr, int64(limit)))
+}
+
+// LimitReader returns a Reader that reads from r
+// but stops with types.ErrLimit after n bytes.
+// The underlying implementation is a *io.LimitedReader.
+func LimitReader(r io.Reader, n int64) io.Reader {
+	return &LimitedReader{r: &io.LimitedReader{R: r, N: n}}
+}
+
+type LimitedReader struct {
+	r *io.LimitedReader
+}
 
-	return ioutil.ReadAll(io.LimitReader(zr, maxSize))
+func (l *LimitedReader) Read(p []byte) (n int, err error) {
+	if l.r.N <= 0 {
+		return 0, types.ErrLimit
+	}
+	return l.r.Read(p)
```

Comment:
ReadAll keeps reading until one read returns `(0, nil)` or any return `(x, err)` correct?

Update: checked [ioutil](https://golang.org/pkg/io/ioutil/#ReadAll) and [bytes.Buffer](https://golang.org/pkg/bytes/#Buffer.ReadFrom). They only stop when it hits an error. If the error is EOF, then it stops and return data with no error, otherwise, it returns the error.

Seems fine, just note this returns an error if the input is the exact size of n.

Reply:
Or maybe, so the read, then check if it overflowed? Or maybe I don't fully understand how LimitedReader works.